src/lib.rs

```diff
@@ -287,69 +327,518 @@ impl<D: AsMut<[Digit]>> BigInt<D> {
+    ///
+    /// It is a low-level function allowing you to manage allocations. Its safe analogue is
+    /// [`&a / b`](#impl-Div%3CDigit%3E).
+    ///
+    /// ## Safety
+    /// Assuming that `a` has [size] `n`:
+    /// * `n > 0`
+    /// * [Size][size] of `self` must be exactly `n` limbs
+    /// * Size of `scratch_space` must be exactly
+    /// [`ScratchSpace::division(n, 1)`](ScratchSpace::division) limbs
+    ///
+    /// [size]: BigInt::size
+    ///
+    /// ## Constant time
+    /// This function is constant time in size of its arguments, ie. number of instructions and
+    /// memory access patterns remain the same for the same `n`.
+    pub unsafe fn div_digit_unchecked(
+        &mut self,
+        a: &mut BigInt<impl AsMut<[Digit]>>,
+        b: Digit,
+        mut scratch_space: ScratchSpace,
+    ) {
+        self.div_unchecked(a, &BigInt::from_digits(&[b]), scratch_space);
     }
 }
 
-impl<'n, D: AsRef<[Digit]>> Add<&'n BigInt<D>> for u16 {
+impl_digit_operation!(Div, div, u8);
+impl_digit_operation_reverse!(Div, div, u8);
+impl_digit_operation!(Div, div, u16);
+impl_digit_operation_reverse!(Div, div, u16);
+impl_digit_operation!(Div, div, u32);
+impl_digit_operation_reverse!(Div, div, u32);
+impl_u64_operation!(Div, div);
+impl_digit_operation_reverse!(Div, div, u64);
+impl_digit_operation_reverse!(Div, div, Digit);
+
+impl<D1, D2> Rem<&BigInt<D2>> for &BigInt<D1>
+where
+    D1: AsRef<[Digit]>,
+    D2: AsRef<[Digit]>,
+{
     type Output = BigInt;
 
-    #[inline(always)]
-    fn add(self, rhs: &BigInt<D>) -> Self::Output {
-        rhs + self
+    /// Computes `self mod rhs`. In other words, the remainder of dividing `self` by `rhs`.
+    ///
+    /// ## Constant time
+    /// Division with remainder is constant time in size of its arguments.
+    /// It means that the number of instructions and memory access patterns remain the same for the same
+    /// `self.size()` and `rhs.size()` (provided that `rhs` does not start from 0).
+    ///
+    /// ## Zeroizing
+    /// This function might allocate some auxiliary buffer on the heap.
+    /// Before deallocating, it gets zeroed out.
+    ///
+    /// ## Complexity
+    /// The complexity is always `O(self.size() * rhs.size())`.
+    ///
+    /// ## Panics
+    /// The method panics if `rhs` equals zero.
+    fn rem(self, rhs: &BigInt<D2>) -> Self::Output {
+        let n = self.size();
+
+        let b = rhs.as_ref();
+        // m is the length of b without leading zeroes
+        let m = ops::len_without_leading_zeroes(b);
+
+        if m == 0 {
+            panic!("Division by zero");
+        }
+
+        // if m > n, then definitely b > a and so a mod b = a.
+        if m > n {
+            return BigInt::from_digits(self.as_ref()).reallocate();
+        }
+
+        // we have to clone `self` in order to be able to use unchecked division with remainder,
+        // as it requires its first argument to be mutable
+        let mut a_copy = vec![Digit::zero(); n];
```

Comment:
Why not `let a_copy = self.reallocate();`?

Reply:
I kind of need a vector of digits there and I'm not sure how to get it after `reallocate`...

Reply:
Ah I see!